packages/hardhat-core/src/internal/hardhat-network/provider/RethnetState.ts

```diff
@@ -0,0 +1,173 @@
+import {
+  Account,
+  Address,
+  bufferToBigInt,
+  KECCAK256_NULL,
+  toBuffer,
+} from "@nomicfoundation/ethereumjs-util";
+import { StateManager, AccountData } from "rethnet-evm";
+import { GenesisAccount } from "./node-types";
+
+/* eslint-disable @nomiclabs/hardhat-internal-rules/only-hardhat-error */
+/* eslint-disable @typescript-eslint/no-unused-vars */
+
+export class RethnetStateManager {
+  constructor(private _state: StateManager = new StateManager()) {}
+
+  public static withGenesisAccounts(
+    genesisAccounts: GenesisAccount[]
+  ): RethnetStateManager {
+    return new RethnetStateManager(
+      StateManager.withGenesisAccounts(
+        genesisAccounts.map((account) => {
+          return {
+            privateKey: account.privateKey,
+            balance: BigInt(account.balance),
+          };
+        })
+      )
+    );
+  }
+
+  public asInner(): StateManager {
+    return this._state;
+  }
+
+  public async accountExists(address: Address): Promise<boolean> {
+    const account = await this._state.getAccountByAddress(address.buf);
+    return account !== null;
+  }
+
+  public async getAccount(address: Address): Promise<Account> {
+    const account = await this._state.getAccountByAddress(address.buf);
+    return new Account(
+      account?.nonce,
+      account?.balance,
+      undefined,
+      account?.codeHash
+    );
+  }
+
+  public async putAccount(address: Address, account: Account): Promise<void> {
+    await this._state.insertAccount(address.buf, {
+      balance: account.balance,
+      nonce: account.nonce,
+      codeHash: account.codeHash,
+    });
+  }
+
+  public async accountIsEmpty(address: Address): Promise<boolean> {
+    const account = await this._state.getAccountByAddress(address.buf);
+    return (
+      account !== null &&
+      account.balance === 0n &&
+      account.nonce === 0n &&
+      account.codeHash.equals(KECCAK256_NULL)
+    );
```

Comment:
Why is this not `return (account === null || (balance === 0 && etc))`?

Reply:
You're right. Wrong application of negation to get rid of the `||` 😅 

Thanks for catching that!